### DISCUSSION TITLE
Merge Database & Podcast Database?

### DISCUSSION THREAD

**Discussion by hasezoey** [tramhao/termusic#291] [Ideas]
Should we in the future maybe merge these 2 databases or should they stay separate?

**Comment by tramhao** [comment]
Perhaps it could save some memory. I think merge them will be better.

**Reply by hasezoey** [reply]
i thought on starting to work on this, but wanted feedback with the database design before committing to it, what do you think about [this database design](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql)?

Open questions:
- [`albums.title`](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L59): TODO: should we force "NOT NULL"?
- [`files_album`](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L73): TODO: should this maybe get integrated to the "files" table as a file can really only belong to one album?
- [Question 1](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L125): should podcast author and music artist link to the same place or be separate?
- [Question 2](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L126): should "created_at" and "modified_at" for local music files also be stored for quick search?
- [Question 3](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L127): should "files.file" be split to "directory, file_stem, ext" instead of just the full path? (similar to the current library.db design)
- [Question 4](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L128): should we do some kind of design where there is one single "last_position" table for both music and podcasts?

anything i missed or would be good to have now?

**Reply by hasezoey** [reply]
@tramhao, could you take a look at this if this is what we should go for?

**Reply by tramhao** [reply]
> i thought on starting to work on this, but wanted feedback with the database design before committing to it, what do you think about [this database design](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql)?
> 
> Open questions:
> 
> * [`albums.title`](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L59): TODO: should we force "NOT NULL"?
Album title should not be forced "NOT NULL" I think.
> * [`files_album`](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L73): TODO: should this maybe get integrated to the "files" table as a file can really only belong to one album?
Yes, there shouldn't be a separate files_album. 
> * [Question 1](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L125): should podcast author and music artist link to the same place or be separate?
I think they should be the same.
> * [Question 2](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L126): should "created_at" and "modified_at" for local music files also be stored for quick search?
These two are not necessary for music. They're from file, not meta data.
> * [Question 3](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L127): should "files.file" be split to "directory, file_stem, ext" instead of just the full path? (similar to the current library.db design)
Sounds better to split.
> * [Question 4](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L128): should we do some kind of design where there is one single "last_position" table for both music and podcasts?

If databases are merged, one single last_position is better.
> 
> anything i missed or would be good to have now?

**Reply by hasezoey** [reply]
> [Question 3](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L127): should "files.file" be split to "directory, file_stem, ext" instead of just the full path? (similar to the current library.db design)

>  Sounds better to split.

I have looked through the code again, and `directory` and `ext` are basically unused and only `name`(`file_stem`) is used, but not for searching and could easily be replaced by just getting it from the full path.
What would we actually do with the split parts in the future? (if its directly search, wouldnt it be better to just do a readdir? and for extension search, should we offer something like that?)
The only thing i could think of this being useful is having the file_stem (or just full filename with extension?) as a fallback to searching for just a metadata title (which could be empty), but if we do this, wouldnt it be enough to split it into `directory + filename` (the filename is file_stem+extension, ie `music.mp3`).

> [Question 4](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L128): should we do some kind of design where there is one single "last_position" table for both music and podcasts?

> If databases are merged, one single last_position is better.

I thought about this a bit, if we do one single table, how should we link them to the files? Have each their own link table (`last_position_files(file, last_position_id)` & `last_position_podcasts(file, last_position_id` & `last_position()`) or have one table with different fields (`last_position(file, podcast)`)?
The more i think about it again, its actually sounds better to just have the `last_position` where it currently is.

---

[Here is the current state with all the updates](https://github.com/hasezoey/termusic/blob/3c2d8d0f94a38b052a77346eb0c5501a4db545cb/lib/design.sql) (except `Question 3` & `Question 4`)

Also another new question as i dont remember exactly why i did this, should we merge `files` & `files_metadata`? (i think the original reason was to split "necessary" from "nullable" / non-necessary metadata)

**Reply by tramhao** [reply]
> > [Question 3](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L127): should "files.file" be split to "directory, file_stem, ext" instead of just the full path? (similar to the current library.db design)
> 
> > Sounds better to split.
> 
> I have looked through the code again, and `directory` and `ext` are basically unused and only `name`(`file_stem`) is used, but not for searching and could easily be replaced by just getting it from the full path. What would we actually do with the split parts in the future? (if its directly search, wouldnt it be better to just do a readdir? and for extension search, should we offer something like that?) The only thing i could think of this being useful is having the file_stem (or just full filename with extension?) as a fallback to searching for just a metadata title (which could be empty), but if we do this, wouldnt it be enough to split it into `directory + filename` (the filename is file_stem+extension, ie `music.mp3`).
In my mind, the database should store original information and later processing could be implemented by share functions. What do you think?
> 
> > [Question 4](https://github.com/hasezoey/termusic/blob/10a80b5ba0826bc1ce13ba1473457357a3ff79f8/lib/design.sql#L128): should we do some kind of design where there is one single "last_position" table for both music and podcasts?
> 
> > If databases are merged, one single last_position is better.
> 
> I thought about this a bit, if we do one single table, how should we link them to the files? Have each their own link table (`last_position_files(file, last_position_id)` & `last_position_podcasts(file, last_position_id` & `last_position()`) or have one table with different fields (`last_position(file, podcast)`)? The more i think about it again, its actually sounds better to just have the `last_position` where it currently is.

I just found today that when I listen to podcasts, the last_position is not remembered(or wrongly matched). Haven't look into it. Perhaps leave them separate is easier for debugging?

> 
> [Here is the current state with all the updates](https://github.com/hasezoey/termusic/blob/3c2d8d0f94a38b052a77346eb0c5501a4db545cb/lib/design.sql) (except `Question 3` & `Question 4`)
> 
> Also another new question as i dont remember exactly why i did this, should we merge `files` & `files_metadata`? (i think the original reason was to split "necessary" from "nullable" / non-necessary metadata)

I don't quite remember the usage of files_metadata. Need look into.

**Comment by Refael10ru** [comment]
Any updates about it/ways I can help it to progress?

**Reply by hasezoey** [reply]
I had not advanced it beyond a SQL "design" for now, as i had focused on some other places to understand better where / how the database is used, and maybe to even remove it in some places (by moving stuff to be server-side).
EDIT: see #425 (and continuation, once its finished) for what i meant

**Comment by hasezoey** [comment]
Part One, the re-implementation for the metadata of the tracks has been done in #511.
